site/_posts/2018-02-01-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+---
+title: Discovering a memory leak in ruby:  An Epic(ish) Tale
+author: NickLaMuro
+date: 2018-02-01
+tags: releases announcements
+comments: true
+published: false
+---
+
+
+_**tl;dr;** we found a [memory leak in
+ruby](https://bugs.ruby-lang.org/issues/14372)... and it mostly wasn't our
+fault!_
+
+This is a tale of that 0.00001% of the time where a problem is actually an
+issue in the language, and not with your own code base... though, it was a bit
+of that as well.  Hopefully what you get out of this post are some ideas for
+how you can solve some issues like this yourself in the future, since we went
+through about [every][1] [possible][2] [blog][3] [post][4] on the subject...
+and [then][5] [some][6]).
+
+Of note, before we get going:  The intent of this post is more of an overview
+of process used to reduce the problem down to it's smallest part, which was
+the biggest, longest, and most frustrating aspect of the effort that was
+undertaken.  Since I am distilling about 4 months of work (hellish struggle)
+into a single post, there will most likely be some future/targeted posts to
+elaborate on some of the specific tools, strategies, and techniques used to
+find the leak.
+
+
+Step 1:  Reproducing the issue
+------------------------------
+
+So to start out, the first issue that was brought to our attention was that
+there was a leak in our `MetricsCollector` and `MetricsProcessor` workers\*.
+This was a trickle leak, where we would leak roughly 5-10 MB per hour, but this
+was unbound memory growth, so eventual this would cause the workers to exceed
```

Comment:
eventually

Reply:
Decided to use "over time" here since we use "We eventually" two paragraphs later, but thanks for the catch!